std/algorithm/comparison.d

```diff
@@ -1565,3 +1568,289 @@ unittest
                 2, "two",
                 ));
 }
+
+alias AllocateGC = Flag!"AllocateGC";
+
+private bool checkLength(Range1, Range2)(Range1 r1, Range2 r2)
+    if (isInputRange!Range1 &&
+        isInputRange!Range2 &&
+        !isInfinite!Range1 &&
+        !isInfinite!Range2)
+{
+    // If the two ranges aren't the same length, then they cannot
+    // possibly be permutations
+    static if (hasLength!(Range1) && hasLength!(Range2))
+    {
+        if (r1.length != r2.length)
+        {
+            return false;
+        }
+    }
+    else
+    {
+        // Custom walk length check to reduce the iterations as much as possible
+        while (!r1.empty)
+        {
+           if (r2.empty)
+           {
+              return false;
+           }
+
+           r1.popFront;
+           r2.popFront;
+        }
+
+        if (!r2.empty)
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/**
+Returns $(D true) if all of the elements in $(D r1) appear the same number of times
+in $(D r2). Otherwise, returns $(D false).
+
+This function can allocate if the $(D AllocateGC.yes) flag is passed. This has
+the benefit of have better complexity than the $(D AllocateGC.no) option. However,
+this option is only available for ranges who's equality can be determined via each
+element's $(D toHash) method. If customized equality is needed, then the $(D pred)
+template parameter can be passed, and the function will automatically switch to
+the non-allocating algorithm. See $(D binaryFun) for more details on how to define
+$(D pred).
+
+Non allocating forward range option: $(BIGOH n^2)
+Non allocating forward range option with custom $(D pred): $(BIGOH n^2)
+Allocating forward range option: amortized $(BIGOH r1.length) + $(BIGOH r2.length)
+
+Params:
+    pred = an optional parameter to change how equality is defined
+    allocate_gc = A $(D std.typecons.Flag!"AllocateGC") instance
+    r1 = A finite forward range
+    r2 = A finite forward range
+
+Returns:
+    $(D bool)
+*/
+
+bool isPermutation(AllocateGC allocate_gc = AllocateGC.no, Range1, Range2)
+(Range1 r1, Range2 r2)
+    if (isForwardRange!Range1 &&
+        isForwardRange!Range2 &&
+        !isInfinite!Range1 &&
+        !isInfinite!Range2)
```

Comment:
I'm a little confused by your suggestions here, because

1. You say `isPermutation(r1, r2)` is ambiguous but you list it in your examples of what should compile

2. All of your examples do as you ask in the current code.